File: 20170628 R Lang/main_Jun01T1.R
# R Program for Task 1 of Jun 01, 2017
# edit in Jun, 28, 2017
# June 1,2017
# May 22-24, 2017
# 20170411 PM
# Mar18,2017
# Task #20170221
# Related to May16/Oct17,2016
# edited in Feb21,2017
rm(list=ls())
setwd("/Users/huwei/Dropbox/temp/20170628 R Lang")
#path0<-getwd()
#if (!is.null(path0)) setwd(path0)
# install.packages("nleqslv")
library("nleqslv")
# use data frame to save and call parameters value
value = c(1.0,0.2,0.6,0.06,0.72,0.1,0.36,0.012,0.06,0.53,1.355,18.0,25.0,0.08)
orgPara = data.frame(value)
row.names(orgPara)=c("pstar","b","phi","sigma","beta","lambda","c_f","r","c_p","delta","A","B1","B2","mu")
#         value
# pstar   1.000
# b       0.200
# phi     0.600
# sigma   0.060
# beta    0.720
# lambda  0.100
# c_f     0.360
# r       0.012
# c_p     0.060
# delta   0.530
# A       1.355
# B1     18.000
# B2     25.000
# mu      0.080
# call the value:
# myPara["pstar","value"]

# variable vector for different cases
lvt <- list(pstar = seq(0.18,2.5,by=0.01),
          b = seq(0.01,1.0,by=0.01),
          phi = seq(0.3,0.95,by=0.01),
          sigma = seq(0.01,1.2,by=0.01),
          beta = seq(0.4,0.99,by=0.01),
          lambda = seq(0.02,0.63,by=0.01),
          c_f = seq(0.12,1.3,by=0.01),
          r = seq(0.003,0.15,by=0.001),
          c_p = seq(0.01,0.2,by=0.001),
          delta = seq(0.21,0.7,by=0.01))
# extract part of object
# example 
# test = "delta"
# vt = lvt[[test]]
########
fun_q_theta <- function(myPara, x) {
   myPara["A","value"]* x **(-myPara["B1","value"]/myPara["B2","value"])
}
## if you want the so-called 'error function'
erf <- function(x) 2 * pnorm(x * sqrt(2)) - 1
## (see Abramowitz and Stegun 29.2.29)
## and the so-called 'complementary error function'

#######################
taskcode<-"Jun01T1"
num_para<-"V27"
str_para0<-paste("paraJun01",num_para,sep = "_")

cell_typen<- c("I","II","III","IV","O")
# 10 cases
cell_case<-c("pstar","b","phi","sigma","beta","lambda","c_f","r","c_p","delta")

v_tt <- 3 # 1:length(cell_typen)
v_cc <- (10)

tt = v_tt
typen=cell_typen[tt]
###############################
for (cc in v_cc){
  casecode=cell_case[cc]
  
  if (typen == "I") {
    epsilon_u <- sqrt(3)
  } else if (typen == "II") {
    epsilon_u <- sqrt(2)/(sqrt(pi-2))
  } else if (typen == "III") {
    epsilon_u <- 1
  } else if (typen == "IV") {
    epsilon_u <- (3-sqrt(3))/2
  } else if (typen == "O") {
    epsilon_u <- (pi/4)/(sin(pi/4))
  }
  
  ### variable vector
  vt = lvt[[casecode]]
  
  str_fig = paste(taskcode,typen,casecode,sep="_")
  
  epsilon_d <-vector()
  epsilon_c <-vector()
  theta <-vector()
  alpha <-vector()
  error<-vector()
  flag<-vector()
  
  for (i in 1:length(vt)) {
    myPara<-orgPara
    myPara[casecode,"value"]<-vt[i]
    # define some functions
    # define F(x)
    fun_F_x <- function(x) {
      if (typen == "O") {
        temp1 = (pi/4.0)/sin(pi/4.0) - x
        temp2 = (pi/2.0)/sin(pi/2.0) - ((pi/4.0)/sin(pi/4.0))**2.0
        temp3 = 1.0 + (temp1/sqrt(temp2))**(-4.0)
        return(1.0 - (temp3)**(-1.0))
      } else if (typen == "I"){
        if(x < -1.0*epsilon_u){
          return(0)
        } else if (x > epsilon_u){
          return(1.0)
        } else {
          return((x + epsilon_u)/(2.0*epsilon_u))}
      } else if (typen == "II") {
        temp=sqrt(1.0/pi)-x * sqrt(.5-1.0/pi)
        return(1.0-erf(temp))
      } else if (typen == 'III') {
        return(.5-.5 * erf((log(-x+1.0)+.5*log(2.0))/sqrt(2.0*log(2.0))))
      } else if (typen == "IV") {
        return (((3.0-2.0*x)/np.sqrt(3.0))**(-3.0))
      }
    }
    # integrate of F(x)
    fun_int_F<-function(a,b) {
      integrand <- function(x) {1.0-fun_F_x(x)}
      temp = integrate(integrand,lower = a, upper = b)
      return(temp$value)
    }
    
    # Nonlinear Equations to be solved
    Equation<-function(x){
      y<-numeric(length(x))
      
      epsilon_d <- x[1]
      epsilon_c <- x[2]
      theta <- x[3]
      alpha <- x[4] 
      
      q_theta = fun_q_theta(myPara=myPara,x=theta)
      
      int_Fdu = fun_int_F(epsilon_d,epsilon_u)
      int_Fcu = fun_int_F(epsilon_c,epsilon_u)
      
      part11 = epsilon_d+myPara["lambda","value"]*int_Fdu/(myPara["r","value"]+myPara["lambda","value"]+theta*q_theta)
      part12 = (myPara["b","value"]-myPara["pstar","value"])/myPara["sigma","value"]
      
      temp211 = myPara['delta',"value"]*(myPara['r',"value"]+myPara['lambda',"value"]-myPara['phi',"value"]*(1-fun_F_x(epsilon_c)))*(epsilon_c-epsilon_d)
      temp212 = myPara['r',"value"]+myPara['lambda',"value"]+theta*q_theta
      temp213 = (myPara['lambda',"value"]/(myPara['r',"value"]+myPara['lambda',"value"])-myPara['lambda',"value"]*myPara['delta',"value"]/temp212)*int_Fcu
      part21 = epsilon_c-temp211/((1-myPara['phi',"value"])*temp212)+temp213
      
      temp2211 = ((myPara['beta',"value"]+myPara['phi',"value"]*(1.0-myPara['beta',"value"]))*myPara['c_f',"value"]*(1.0-alpha))/((1-myPara['beta',"value"])*(1-myPara['phi',"value"]))
      temp2212 = (myPara['beta',"value"]*myPara['c_p',"value"]*alpha)/(1-myPara['beta',"value"])
      temp221 = theta*(temp2211 + temp2212)
      part22 = myPara['delta',"value"]*epsilon_d+((1-myPara['delta',"value"])*(myPara['b',"value"]-myPara['pstar',"value"])+temp221)/myPara['delta',"value"]
      
      part31 = 1/q_theta
      
      temp321 = (1-myPara['beta',"value"])*(1-myPara['phi',"value"])/myPara['c_f',"value"]
      temp322 = myPara['sigma',"value"]*(epsilon_u-epsilon_c)/(myPara['r',"value"]+myPara['lambda',"value"])+myPara['delta',"value"]*myPara['sigma',"value"]*(epsilon_c-epsilon_d)/((1-myPara['phi',"value"])*(myPara['r',"value"]+myPara['lambda',"value"]+theta*q_theta))
      part32 = temp321*temp322
      
      part41 = 1/q_theta
      
      temp421 = (1-myPara['beta',"value"])*myPara['delta',"value"]*myPara['sigma',"value"]*(epsilon_u-epsilon_d)
      temp422 = myPara['c_p',"value"]*(myPara['r',"value"]+myPara['lambda',"value"]+theta*q_theta)
      part42 = temp421/temp422
      
      y[1]<-part11-part12
      y[2]<-part21-part22
      y[3]<-part31-part32
      y[4]<-part41-part42
      
      y
    }
    # operator to solve the nonlinear equations
    
    # A numeric vector with an initial guess of the root of the function.
    xstart<-c(-5.0,-1.0,1.0,0.9)
    # trace: trace :Non-negative integer. A value of 1 will give a detailed report of the progress of the iteration.
    
    sol<-nleqslv(xstart,Equation,method="Broyden",control = list(trace=1))
    
    epsilon_d[i]<-sol$x[1]
    epsilon_c[i]<-sol$x[2]
    theta[i]<-sol$x[3]
    alpha[i]<-sol$x[4]
    
    # Euclidean norm of function values
    error[i]<-norm(sol$fvec,type="2")
    # termination code as integer. The values returned are:
    # 1
    # Function criterion is near zero. Convergence of function values has been achieved.
    # 2
    # x-values within tolerance. This means that the relative distance between two consecutive x-values is smaller than xtol but that the function value criterion is still larger than ftol. Function values may not be near zero; therefore the user must check if function values are acceptably small.
    # 3
    # No better point found. This means that the algorithm has stalled and cannot find an acceptable new point. This may or may not indicate acceptably small function values.
    # 4
    # Iteration limit maxit exceeded.
    # 5
    # Jacobian is too ill-conditioned.
    # 6
    # Jacobian is singular.
    # 7
    # Jacobian is unusable.
    # -10
    # User supplied Jacobian is most likely incorrect.
    flag[i]<- sol$termcd
  }
  # Data Outpu
  write.table(epsilon_d,file=paste(str_fig,str_para0,"epsilon_d.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
  # temp<-read.table(file=paste(str_fig,str_para0,"epsilon_d.txt",sep="_"),sep="\n")
  # epsilon_d = temp[[1]]
  write.table(epsilon_c,file=paste(str_fig,str_para0,"epsilon_c.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
  write.table(theta,file=paste(str_fig,str_para0,"theta.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
  write.table(alpha,file=paste(str_fig,str_para0,"alpha.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
  write.table(error,file=paste(str_fig,str_para0,"error.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
  write.table(flag,file=paste(str_fig,str_para0,"flag.txt",sep="_"),sep="\n",row.name=FALSE,col.names = FALSE)
}

File: 4 Python/main_Jun01T1.py
# main_Jun01T1.py

# Python program for Task 1 of Jun 01,2017
# June 1,2017
# May 22-24, 2017
# 20170411 PM
# Mar18,2017
# Task #20170221
# Related to May16/Oct17,2016
# edited in Feb21,2017
import numpy as np
import numpy.linalg as LA
import scipy.special as special
import scipy.integrate as integrate
from scipy.optimize import fsolve
import os

#
class Epsilon_u:
	"""class of epsilon_u"""

	def __init__(self, typen):
		# self.type is object variable
		self.typen = typen

	def Assign(self):
		if self.typen == 'I':
			return np.sqrt(3)
		elif self.typen == 'II':
			return np.sqrt(2)/(np.sqrt(np.pi-2))
		elif self.typen == 'III':
			return 1
		elif self.typen == 'IV':
			return (3-np.sqrt(3))/2
		elif self.typen == 'O':
			return (np.pi/4)/(np.sin(np.pi/4))

#
class CaseCode:
	"""docstring for ClassName"""
	def __init__(self, case):
		self.case = case

	def VT(self):
		# 1
		if self.case == 'pstar':
			return [0.01*x for x in range(18,251,1)]
		# 2
		elif self.case == 'b':
			return [0.01*x for x in range(1,101,1)]
		# 3
		elif self.case == 'phi':
			return [0.01*x for x in range(30,96,1)]
		# 4
		elif self.case == 'sigma':
			return [0.01*x for x in range(1,121,1)]
		# 5
		elif self.case == 'beta':
			return [0.01*x for x in range(40,100,1)]
		# 6
		elif self.case == 'lambda':
			return [0.01*x for x in range(2,64,1)]
		# 7
		elif self.case == 'c_f':
			return [0.01*x for x in range(12,131,1)]
		# 8
		elif self.case == 'r':
			return [0.001*x for x in range(3,151,1)]
		# 9
		elif self.case == 'c_p':
			return [0.001*x for x in range(10,201,1)]
		# 10
		elif self.case == 'delta':
			return [0.01*x for x in range(21,71,1)]

	def Str_Var(self):
		"""LaTex string"""
		# 1
		if self.case == 'pstar':
			return r'p^*'
		# 2
		elif self.case == 'b':
			return r'b'
		# 3
		elif self.case == 'phi':
			return r'\phi'
		# 4
		elif self.case == 'sigma':
			return r'\sigma'
		# 5
		elif self.case == 'beta':
			return r'\beta'
		# 6
		elif self.case == 'lambda':
			return r'\lambda'
		# 7
		elif self.case == 'c_f':
			return r'c^F'
		# 8
		elif self.case == 'r':
			return r'r'
		# 9
		elif self.case == 'c_p':
			return r'c^P'
		# 10
		elif self.case == 'delta':
			return r'\delta'

	def Tell(self):
		print('Case is: {}'.format(self.case))
		print('The array of variable is:{}'.format(CaseCode.VT(self)))
		print('String of variable is:{}'.format(CaseCode.Str_Var(self)))

#
class Operator(CaseCode):
	"""Main operator to get the solution

	epsilon_d,epsilon_c,theta,alpha"""

	# para is the dictionary updated
	def __init__(self, case, typen, para):
		CaseCode.__init__(self, case)
		self.typen = typen
		self.para = para

	def fun_q_theta(self, x):
		return self.para['A']* x **(-self.para['B1']/self.para['B2'])

	def fun_F_x(self, x):
		if self.typen == 'O':
			temp1 = (np.pi/4.0)/np.sin(np.pi/4.0) - x
			temp2 = (np.pi/2.0)/np.sin(np.pi/2.0) - ((np.pi/4.0)/np.sin(np.pi/4.0))**2.0
			temp3 = 1.0 + (temp1/np.sqrt(temp2))**(-4.0)
			return 1.0-(temp3)**(-1.0)
		elif self.typen == 'I':
			if x < -1.0*self.para['epsilon_u']:
				return 0
			elif x > self.para['epsilon_u']:
				return 1.0
			else:
				return (x + self.para['epsilon_u'])/(2.0*self.para['epsilon_u'])

		elif self.typen == 'II':
			temp=np.sqrt(1.0/np.pi)-x*np.sqrt(.5-1.0/np.pi)
			return 1.0-special.erf(temp)

		elif self.typen == 'III':
			return (.5-.5 * special.erf((np.log(-x+1.0)+ \
				.5 * np.log(2.0))/np.sqrt(2.0 * np.log(2.0))))

		elif self.typen == 'IV':
			return ((3.0-2.0*x)/np.sqrt(3.0))**(-3.0)

	def integrand(self,x):
		return (1-self.fun_F_x(x))

	def fun_int_F(self,a,b):
		temp = integrate.quad(self.integrand,a,b)
		return temp[0]

	def Equation(self,x):
		epsilon_d = x[0]
		epsilon_c = x[1]
		theta = x[2]
		alpha = x[3]

		q_theta = self.fun_q_theta(theta)

		int_Fdu = self.fun_int_F(epsilon_d,self.para['epsilon_u'])
		int_Fcu = self.fun_int_F(epsilon_c,self.para['epsilon_u'])
		####
		part11 = epsilon_d + self.para['lambda'] * \
			int_Fdu / (self.para['r'] + self.para['lambda'] + theta * q_theta)
		part12 = (self.para['b'] - self.para['pstar'])/self.para['sigma']
		####
		temp211 = self.para['delta'] * (self.para['r'] + self.para['lambda']\
			-self.para['phi'] * (1.0-self.fun_F_x(epsilon_c)))\
				 * (epsilon_c-epsilon_d)
		temp212 = self.para['r'] + self.para['lambda'] + theta * q_theta
		temp213 = (self.para['lambda']/(self.para['r']+self.para['lambda'])\
			-self.para['lambda'] * self.para['delta']/temp212) * int_Fcu
		part21 = epsilon_c-temp211/((1.0-self.para['phi']) * temp212)+temp213

		temp2211 = ((self.para['beta']+self.para['phi'] * \
			(1.0-self.para['beta'])) * self.para['c_f'] * (1.0-alpha))/((1.0-self.para['beta']) * \
			(1.0-self.para['phi']))
		temp2212 = (self.para['beta'] * self.para['c_p'] * alpha)/(1.0-self.para['beta'])
		temp221 = theta * (temp2211 + temp2212)
		part22 = self.para['delta'] * epsilon_d+((1.0-self.para['delta']) * \
			(self.para['b']-self.para['pstar'])+temp221)/self.para['delta']
		####
		part31 = 1.0/q_theta

		temp321 = (1.0-self.para['beta']) * (1.0-self.para['phi'])/self.para['c_f']
		temp322 = self.para['sigma'] * (self.para['epsilon_u']-epsilon_c)/(\
			self.para['r']+self.para['lambda'])+\
			self.para['delta'] * self.para['sigma'] * (epsilon_c-epsilon_d)/(\
				(1.0-self.para['phi']) * (self.para['r']+self.para['lambda']+theta * q_theta))
		part32 = temp321 * temp322
		####
		part41 = 1.0/q_theta

		temp421 = (1.0-self.para['beta']) * self.para['delta'] * self.para['sigma'] * (\
			self.para['epsilon_u']-epsilon_d)
		temp422 = self.para['c_p'] * (self.para['r']+self.para['lambda']+theta * q_theta)
		part42 = temp421/temp422

		return [
			part11-part12,
			part21-part22,
			part31-part32,
			part41-part42
		]
	def Solution(self):
		# Tuple
		# Initial value guess
		guess = (-5.0,-1.0,1.0,0.5)

		# fsolve is a wrapper around MINPACK's hybrd and hybrj algorithms
		# [epsilon_d,epsilon_c,theta,alpha]
		# xtol : The calculation will terminate if the relative error between
		# two consecutive iterates is at most xtol.
		# maxfev : The maximum number of calls to the function.
		sol, infodict, ier, mesg = fsolve(self.Equation,guess,full_output=True)


		if ier != 1:
			print("ier = {}".format(ier))
			print("sol = {}".format(sol))
			print("infodict = {}".format(infodict))
			return [sol,ier]
		else:
		 	return [sol,ier]


	def Tell(self):
		CaseCode.Tell(self)
		print('Type of F(x) is:{}'.format(self.typen))
		print('Parameters are:{}'.format(self.para))
		print('Solution is:{}'.format(self.Solution))
		print('Error is:{}'.format(self.Error))

# main part
taskcode = 'Jun01T1'
str_para = 'paraJun01V27'
# Example on Windows:
# source = ['"C:\\My Documents"','C:\\Code']
# for Mac OS or Linux
path_data = '/Users/huwei/Dropbox/On_Local/Python/20170620 Python/'
if not os.path.exists(path_data):
	print('path_data does not exist!')
# casecode: a list
cell_typen = ['I','II','III','IV','O']
cell_case = ['pstar','b','phi','sigma','beta','lambda',\
'c_f','r','c_p','delta']

v_tt=[3]
v_cc=range(1,11)

# loop for type of function
for tt in v_tt:
	typen = cell_typen[tt-1]
	print(typen)
	# loop for case
	for cc in v_cc:
		# case code
		case = cell_case[cc-1]

		myParameter_Jun01V27 = {
			'pstar': 1,
			'b': 0.2,
			'phi': 0.6,
			'sigma': 0.06,
			'beta': 0.72,
			'lambda': 0.1,
			'c_f': 0.36,
			'r': 0.012,
			'c_p': 0.06,
			'delta': 0.53,

			'A': 1.355,
			'B1': 18.0,
			'B2': 25.0,
			'mu': 0.08,
			'epsilon_u':Epsilon_u(typen).Assign()
		}


		# array of variable
		vt=CaseCode(case).VT()
		str_fig = taskcode+'_'+typen+'_'+case

		vepsilon_d=[]
		vepsilon_c=[]
		vtheta=[]
		valpha=[]
		vflag=[]

		for i in range(0,len(vt)):
			# initialize parameters in each loop
			Paras = myParameter_Jun01V27
			Paras[case] = vt[i]
			tresult = Operator(case,typen,Paras).Solution()
			vepsilon_d.append(tresult[0][0])
			vepsilon_c.append(tresult[0][1])
			vtheta.append(tresult[0][2])
			valpha.append(tresult[0][3])
			vflag.append(tresult[1])

		np.savetxt(path_data+os.sep+str_fig+'_'+str_para+'_epsilon_d.txt',vepsilon_d)
		np.savetxt(path_data+os.sep+str_fig+'_'+str_para+'_epsilon_c.txt',vepsilon_c)
		np.savetxt(path_data+os.sep+str_fig+'_'+str_para+'_theta.txt',vtheta)
		np.savetxt(path_data+os.sep+str_fig+'_'+str_para+'_alpha.txt',valpha)
		np.savetxt(path_data+os.sep+str_fig+'_'+str_para+'_flag.txt',vflag)

		################################################
		#print typen
		#print CaseCode(case).VT()[1]
		#print Epsilon_u(typen).Assign()
		#print myParameter_Jun01V27['pstar']
		#basic
